src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/UntypedReferenceRewriter.cs

```diff
@@ -0,0 +1,112 @@
+{
+    /// <summary>
+    /// For expressions over a reference search parameter where the type is not specified
+    /// (e.g. "abc" instead of "Patient/123") AND the search parameter is constrained to be
+    /// of only one type, rewrites the expression to specify the single target type.
+    /// </summary>
+    internal class UntypedReferenceRewriter : ExpressionRewriterWithInitialContext<object>
+    {
+        public static readonly UntypedReferenceRewriter Instance = new UntypedReferenceRewriter();
+
+        public override Expression VisitSearchParameter(SearchParameterExpression expression, object context)
+        {
+            // Handle reference search parameters as well as composite search parameters with one or more reference components.
+            // We first create a bitmask with bits set representing the component indexes that are candidates for this rule.
+            // Bit 0 is for reference, non-composite parameters where the reference is of a single possible type
+            // Bits 1 and up represent the component indexes (plus one) where the component is a reference search parameter with one type
+
+            int componentCandidates = expression.Parameter.Type switch
+            {
+                SearchParamType.Reference when expression.Parameter.TargetResourceTypes?.Count == 1 => 1,
+                SearchParamType.Composite =>
+                    expression.Parameter.ResolvedComponents.Aggregate(
+                            (index: 1, flags: 0),
+                            (acc, p) => (index: acc.index + 1, flags: acc.flags | (p.Type == SearchParamType.Reference && p.TargetResourceTypes?.Count == 1 ? 1 << acc.index : 0)))
+                        .flags,
+                _ => 0
+            };
+
+            if (componentCandidates == 0)
+            {
+                // nothing to do for this expression
+                return expression;
+            }
+
+            // now see which components have a expression on the reference type
+            int componentsPresent = expression.Expression.AcceptVisitor(ParameterPredicateVisitor.Instance, null);
+
+            // Now determine which components should get a type expression added to it.
+            // componentCandidates will have bits set for each component that we could provide a known type expression for.
+            // componentsPresent will have bits set for each component that actually has a type expression.
+            // So the components that we need to provide a type expression for can be obtained by a set difference,
+            // which we can do with a bitwise complement (~) and bitwise intersection (&).
+
+            int componentsToFill = componentCandidates & ~componentsPresent;
+
+            if (componentsToFill == 0)
+            {
+                return expression;
+            }
+
+            List<Expression> newExpressionsToBeAnded;
+            if (expression.Expression is MultiaryExpression me)
```

Comment:
>MultiaryExpression me [](start = 41, length = 21)

Would it worth to add some Ensure or Debug.Assert here what me MultiaryOperation is actually And?
Could go terribly wrong if we AND an OR expression (which I assume never the case because that's not how code operates right now, but i think it's better be safe.

Reply:
Good idea

Reply:
...And in so doing I completely missed ORed expressions like Patient?Organization=a,b 
😱 
Fixed that now. Thanks for the comment. I added the check to `ParameterPredicateVisitor.VisitMultiary()`